### PR TITLE
Intake v2 transport

### LINKED ISSUE #204: Update span_count values and keys
https://github.com/elastic/apm-server/issues/1241#issuecomment-419977863

### PR DESCRIPTION
**Pull request by mikker**
Closes elastic/apm-agent-ruby#181 

Stil todo in this PR:

- [x] ~Cycle~ Close connection when max time reached
- [x] ~Cycle~ Close connection when max bytes
- [x] Add `metadata`
- [x] gzip
- [x] Handle errors?
    1. ~Back off on errors (`min(errorCount++, 6) ** 2`)~ Pushed to later version
    1. ~Buffer requests until recovered~
- [x] Update records according to schema changes
- [x] Make sure it's thread safe

In other PRs:

- Move `submit`ing txn/spans out of the objects themselves
- Update benchmark script

---

Closes #204

### CODE REVIEW

**Comment by mikker** [comment]
jenkins, retest this please

**Comment by mikker** [comment]
@simitt I think this is ready for review. Apologies for the size 😬

Most things are concentrated around the new `Transport` module and especially `Transport::Base` and `Transport::Connection`

File: spec/elastic_apm/transport/base_spec.rb [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+          allow(subject.connection).to receive(:write) { true }
+
+          subject.post(
+            transaction: {
+              id: 1,
+              context: { request: { headers: { ApiKey: 'OH NO!' } } }
+            }
+          )
+
+          expect(subject.connection).to have_received(:write)
```
Comment (simitt) [inline spec/elastic_apm/transport/base_spec.rb]: How does this check that sensitive data are removed?

File: spec/elastic_apm/transport/base_spec.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+          agent.stop
+        end
+
+        it 'starts all requests with a metadata object' do
+          subject.post({})
+          subject.flush
+          subject.post({})
+          subject.flush
+          expect(@mock_intake.requests.length).to be 2
+          expect(@mock_intake.metadatas.length).to be 2
```
Comment (simitt) [inline spec/elastic_apm/transport/base_spec.rb]: Can you add a test checking that multiple events can be written via the same request?

File: spec/support/match_json_schema_matcher.rb [hunk 3]
```diff
@@ -3,8 +3,8 @@
 require 'spec_helper'
 
 SCHEMA_URLS = {
-  transactions: 'https://github.com/elastic/apm-server/raw/master/docs/spec/transactions/payload.json',
-  errors: 'https://github.com/elastic/apm-server/raw/master/docs/spec/errors/payload.json'
+  transactions: 'https://github.com/simitt/apm-server/raw/1276-update-json-spec-v2/docs/spec/transactions/v2_transaction.json',
+  errors: 'https://github.com/simitt/apm-server/raw/1276-update-json-spec-v2/docs/spec/errors/v2_error.json'
```
Comment (simitt) [inline spec/support/match_json_schema_matcher.rb]: The PRs are merged, you should refer to `https://github.com/elastic/apm-server/tree/v2/docs/spec` now.

File: lib/elastic_apm/config.rb [hunk 4]
```diff
@@ -19,12 +19,15 @@ class Config
...
       log_level: Logger::DEBUG,
 
-      max_queue_size: 500,
-      flush_interval: 10,
       transaction_sample_rate: 1.0,
       transaction_max_spans: 500,
       filter_exception_types: [],
 
+      # intake v2
+      api_request_size: 786_432,
```
[the remark this answers is not in the document]
Comment (mikker) [reply lib/elastic_apm/config.rb]: I think I got it from one of the other agents? I'll go see what it's supposed to default to.

File: lib/elastic_apm/span.rb [hunk 5]
```diff
@@ -51,8 +57,13 @@ def done
 
       self.original_backtrace = nil # release it
 
+      if @transaction.instrumenter
+        @transaction.instrumenter.submit_span self # TODO: move this
```
[the remark this answers is not in the document]
Comment (mikker) [reply lib/elastic_apm/span.rb]: Actually future TODO to do it in another PRhttps://github.com/elastic/apm-agent-ruby/pull/192

File: lib/elastic_apm/transport/connection.rb [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            close!
+          end
+      end
+
+      def enable_compression
+        @wr.binmode
+        @wr = Zlib::GzipWriter.new(@wr)
+      end
+
+      def reset!
```
[the remark this answers is not in the document]
Comment (mikker) [reply lib/elastic_apm/transport/connection.rb]: You are right!

File: lib/elastic_apm/error/exception.rb [hunk 7]
```diff
@@ -4,7 +4,7 @@ module ElasticAPM
   class Error
     # @api private
     class Exception
-      MOD_SPLIT = '::'.freeze
```
[the remark this answers is not in the document]
Comment (mikker) [reply lib/elastic_apm/error/exception.rb]: Removed official Ruby 2.2 support in the middle of this PR :/
2.3 added the magic comment `frozen_string_literal` that makes all strings frozen, so no longer necessary to do on each individual string.

File: spec/elastic_apm/transport/base_spec.rb (hunk 1 above)
Comment (mikker) [reply spec/elastic_apm/transport/base_spec.rb]: It doesn't. I suppose there's a filter spec? Do you think I should add that expectation to this spec as well?

Comment (simitt) [reply spec/elastic_apm/transport/base_spec.rb]: Either this, or change the spec description (currently it is `it 'filters sensitive data' do`).

**Comment by mikker** [comment]
Thanks! No, haven't done any benchmarks. But so far these are required changes to work with v2, so fster or not they are necessary. I will do some tests and benchmarks before 2.0 of course.

File: lib/elastic_apm/transport/connection.rb [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+        return unless @bytes_sent >= @config.api_request_size
+
+        close!
+      rescue FailedToConnectError => e
+        error "Couldn't establish connection to APM Server:\n%p", e
+
+        nil
+      end
+
+      def connected?
```
[the remark this answers is not in the document]
Comment (mikker) [reply lib/elastic_apm/transport/connection.rb]: Used in a lot of tests as `expect(...).to be_connected`

**Pull request merged**